// @tamagui/theme-builder/dist/cjs/masks.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var masks_exports = {};
__export(masks_exports, {
  maskOptions: () => maskOptions,
  masks: () => masks
});
module.exports = __toCommonJS(masks_exports);
var import_create_theme = require("@tamagui/create-theme");
const shadows = {
  shadowColor: 0,
  shadowColorHover: 0,
  shadowColorPress: 0,
  shadowColorFocus: 0
}, colors = {
  ...shadows,
  color: 0,
  colorHover: 0,
  colorFocus: 0,
  colorPress: 0
}, templateColorsSpecific = {
  color1: 1,
  color2: 2,
  color3: 3,
  color4: 4,
  color5: 5,
  color6: 6,
  color7: 7,
  color8: 8,
  color9: 9,
  color10: 10,
  color11: 11,
  color12: 12
}, skipShadowsAndSpecificColors = {
  ...shadows,
  ...templateColorsSpecific
}, PALETTE_LEN = 14, baseMaskOptions = {
  override: shadows,
  skip: shadows,
  // avoids the transparent ends
  max: PALETTE_LEN - 2,
  min: 1
}, maskOptions = {
  component: {
    ...baseMaskOptions,
    override: colors,
    skip: skipShadowsAndSpecificColors
  },
  alt: {
    ...baseMaskOptions
  },
  button: {
    ...baseMaskOptions,
    override: {
      ...colors,
      borderColor: "transparent",
      borderColorHover: "transparent"
    },
    skip: skipShadowsAndSpecificColors
  }
}, masks = {
  identity: (0, import_create_theme.createIdentityMask)(),
  soften: (0, import_create_theme.createSoftenMask)(),
  soften2: (0, import_create_theme.createSoftenMask)({ strength: 2 }),
  soften3: (0, import_create_theme.createSoftenMask)({ strength: 3 }),
  strengthen: (0, import_create_theme.createStrengthenMask)(),
  inverse: (0, import_create_theme.createInverseMask)(),
  inverseSoften: (0, import_create_theme.combineMasks)((0, import_create_theme.createInverseMask)(), (0, import_create_theme.createSoftenMask)({ strength: 2 })),
  inverseSoften2: (0, import_create_theme.combineMasks)((0, import_create_theme.createInverseMask)(), (0, import_create_theme.createSoftenMask)({ strength: 3 })),
  inverseSoften3: (0, import_create_theme.combineMasks)((0, import_create_theme.createInverseMask)(), (0, import_create_theme.createSoftenMask)({ strength: 4 })),
  inverseStrengthen2: (0, import_create_theme.combineMasks)(
    (0, import_create_theme.createInverseMask)(),
    (0, import_create_theme.createStrengthenMask)({ strength: 2 })
  ),
  strengthenButSoftenBorder: (0, import_create_theme.createMask)((template, options) => {
    const stronger = (0, import_create_theme.createStrengthenMask)().mask(template, options), softer = (0, import_create_theme.createSoftenMask)().mask(template, options);
    return {
      ...stronger,
      borderColor: softer.borderColor,
      borderColorHover: softer.borderColorHover,
      borderColorPress: softer.borderColorPress,
      borderColorFocus: softer.borderColorFocus
    };
  }),
  soften2Border1: (0, import_create_theme.createMask)((template, options) => {
    const softer2 = (0, import_create_theme.createSoftenMask)({ strength: 2 }).mask(template, options), softer1 = (0, import_create_theme.createSoftenMask)({ strength: 1 }).mask(template, options);
    return {
      ...softer2,
      borderColor: softer1.borderColor,
      borderColorHover: softer1.borderColorHover,
      borderColorPress: softer1.borderColorPress,
      borderColorFocus: softer1.borderColorFocus
    };
  }),
  soften3FlatBorder: (0, import_create_theme.createMask)((template, options) => {
    const borderMask = (0, import_create_theme.createSoftenMask)({ strength: 2 }).mask(template, options);
    return {
      ...(0, import_create_theme.createSoftenMask)({ strength: 3 }).mask(template, options),
      borderColor: borderMask.borderColor,
      borderColorHover: borderMask.borderColorHover,
      borderColorPress: borderMask.borderColorPress,
      borderColorFocus: borderMask.borderColorFocus
    };
  }),
  softenBorder: (0, import_create_theme.createMask)((template, options) => {
    const plain = import_create_theme.skipMask.mask(template, options), softer = (0, import_create_theme.createSoftenMask)().mask(template, options);
    return {
      ...plain,
      borderColor: softer.borderColor,
      borderColorHover: softer.borderColorHover,
      borderColorPress: softer.borderColorPress,
      borderColorFocus: softer.borderColorFocus
    };
  }),
  softenBorder2: (0, import_create_theme.createMask)((template, options) => {
    const plain = import_create_theme.skipMask.mask(template, options), softer = (0, import_create_theme.createSoftenMask)({ strength: 2 }).mask(template, options);
    return {
      ...plain,
      borderColor: softer.borderColor,
      borderColorHover: softer.borderColorHover,
      borderColorPress: softer.borderColorPress,
      borderColorFocus: softer.borderColorFocus
    };
  })
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  maskOptions,
  masks
});
